src/peer/negotiator.js

```diff
@@ -477,6 +438,80 @@ class Negotiator extends EventEmitter {
   }
 
   /**
+   * Replace the stream being sent with a new one.
+   * Video and audio are replaced per track by using `xxxTrack` methods.
+   * We assume that there is at most 1 audio and at most 1 video in local stream.
+   * @param {MediaStream} newStream - The stream to replace the old stream with.
+   * @private
+   */
+  _replacePerTrack(newStream) {
+    const _this = this;
+    const vTracks = newStream.getVideoTracks();
+    const aTracks = newStream.getAudioTracks();
+
+    const senders = this._pc.getSenders();
+    const vSender = senders.find(sender => sender.track.kind === 'video');
+    const aSender = senders.find(sender => sender.track.kind === 'audio');
+
+    _updateSenderWithTrack(vSender, vTracks[0], newStream);
+    _updateSenderWithTrack(aSender, aTracks[0], newStream);
+
+    this._replaceStreamCalled = true;
+
+    // We don't actually need to do renegotiation but force it in order to prevent
+    // problems with the stream.id being mismatched when renegotiation happens anyways
+    // Firefox 55 doesn't trigger onnegotiationneeded event when a track in a stream is removed
+    this._pc.onnegotiationneeded();
+
+    /**
+     * Replace a track being sent with a new one.
+     * @param {RTCRtpSender} sender - The sender which type is video or audio.
+     * @param {MediaStreamTrack} track - The track of new stream.
+     * @param {MediaStream} stream - The stream which contains the track. 
+     * @private
+     */
+    function _updateSenderWithTrack(sender, track, stream) {
+      if (track === undefined && sender === undefined) {
+        return;
+      }
+      // remove video or audio sender if not passed
+      if (track === undefined) {
+        _this._pc.removeTrack(sender);
+        return;
+      }
+      // if passed, replace track or create sender
+      if (sender === undefined) {
+        _this._pc.addTrack(track, stream);
+        return;
+      }
+      // if track was not replaced, do nothing
+      if (sender.track.id === track.id) {
```

Comment:
The tests are failing

![screen shot 2017-10-05 at 16 39 03](https://user-images.githubusercontent.com/2532867/31215873-46f579a2-a9ec-11e7-96ee-dada45d88ec2.png)

I'm not sure but I think this line might be the problem. Could you see what setting the track.id in the tests does?

Reply:
In this test, both stubs of senders and stubs of tracks do not have `id` key. So I think this statement is always true because it is `undefined === undefined`.